src/feature/hs/hs_service.c

```diff
@@ -3404,6 +3404,29 @@ service_encode_descriptor(const hs_service_t *service,
   return ret;
 }
 
+/* An introduction circuit has timed out while building or waiting for an
+ * established introduction cell. */
+static void
+service_intro_circuit_timed_out(const origin_circuit_t *circ)
+{
+  hs_service_t *service = NULL;
+  hs_service_intro_point_t *ip = NULL;
+
+  tor_assert(circ);
+  tor_assert(TO_CIRCUIT(circ)->purpose == CIRCUIT_PURPOSE_S_ESTABLISH_INTRO ||
```

Comment:
Let's ditch the fatal assert just in case something changes in the future?

Reply:
Hmmm well that is the whole purpose of that assert. If we change something later that sends something else than an introduction circuit, we can catch it _early_. Because of the switch case in `hs_service_circuit_timed_out()`, it should never happen so this assert would be hit instantly by our unit testing.

Reply:
I've also added a non fatal assert on `circuit_established` so we could catch more bugs if any.